### DISCUSSION TITLE
r n s h - Shell over Reticulum

### DISCUSSION THREAD

**Discussion by acehoss** [markqvist/Reticulum#225] [Show and tell]
Hello! I wanted to share a little project I've been working on, which I call `rnsh`.

`rnsh` is a utility written in Python that facilitates shell sessions over Reticulum networks. It is based on the `rnx` utility that ships with Reticulum and aims to have provide a similar experience to SSH. `rnsh` is still a little raw; some things are implemented badly, and many other things haven't been built at all (yet).

I've only tested it on a few machines and like a weirdo my primary machines run macOS. But `rnsh` should run on anything Unix-like with Python 3.9+.

`pip3 install rnsh` or download from the [rnsh GitHub repository](https://github.com/acehoss/rnsh). The GitHub page also has some basic instructions for getting started. For a taste, here is a usage example.

_Updated for v0.0.5_

### Setup
Before running the listener or initiator, you'll need to get the 
listener destination hash and the initiator identity hash.
```shell
# On listener
rnsh -l -p

# On initiator
rnsh -p
```

### Listener (Server)
- Listening for default service name ("default").
- Using user's default Reticulum config dir (~/.reticulum).
- Using default identity ($RNSCONFIGDIR/storage/identities/rnsh).
- Allowing remote identity `6d47805065fa470852cf1b1ef417a1ac` to connect.
- On connect, will launch the shell of the user running the process, or run the
   command specified by the initiator, if supplied.
```shell
rnsh -l -a 6d47805065fa470852cf1b1ef417a1ac
```
### Initiator (Client)
- Using user's default Reticulum config dir (~/.reticulum).
- Using default identity ($RNSCONFIGDIR/storage/identities/rnsh).
- Connecting to destination `a5f72aefc2cb3cdba648f73f77c4e887`
```shell
rnsh a5f72aefc2cb3cdba648f73f77c4e887
```

### All Options
```
Usage:
    rnsh -l [-c <configdir>] [-i <identityfile> | -s <service_name>] [-v... | -q...] -p
    rnsh -l [-c <configdir>] [-i <identityfile> | -s <service_name>] [-v... | -q...] 
            [-b <period>] (-n | -a <identity_hash> [-a <identity_hash>] ...) [-A | -C] 
            [[--] <program> [<arg> ...]]
    rnsh [-c <configdir>] [-i <identityfile>] [-v... | -q...] -p
    rnsh [-c <configdir>] [-i <identityfile>] [-v... | -q...] [-N] [-m] [-w <timeout>] 
         <destination_hash> [[--] <program> [<arg> ...]]
    rnsh -h
    rnsh --version

Options:
    -c DIR --config DIR          Alternate Reticulum config directory to use
    -i FILE --identity FILE      Specific identity file to use
    -s NAME --service NAME       Service name for identity file if not default
    -p --print-identity          Print identity information and exit
    -l --listen                  Listen (server) mode. If supplied, <program> <arg>...will 
                                   be used as the command line when the initiator does not
                                   provide one or when remote command is disabled. If
                                   <program> is not supplied, the default shell of the 
                                   user rnsh is running under will be used.
    -b --announce PERIOD         Announce on startup and every PERIOD seconds
                                 Specify 0 for PERIOD to announce on startup only.
    -a HASH --allowed HASH       Specify identities allowed to connect
    -n --no-auth                 Disable authentication
    -N --no-id                   Disable identify on connect
    -A --remote-command-as-args  Concatenate remote command to argument list of <program>/shell
    -C --no-remote-command       Disable executing command line from remote
    -m --mirror                  Client returns with code of remote process
    -w TIME --timeout TIME       Specify client connect and request timeout in seconds
    -q --quiet                   Increase quietness (move level up), multiple increases effect
                                          DEFAULT LOGGING LEVEL
                                                  CRITICAL (silent)
                                    Initiator ->  ERROR
                                                  WARNING
                                     Listener ->  INFO
                                                  DEBUG    (insane)
    -v --verbose                 Increase verbosity (move level down), multiple increases effect
    --version                    Show version
    -h --help                    Show this help
```
### Updated 2023-02-22 for v0.0.12

**Comment by markqvist** [comment]
This is so cool @acehoss. I'm flabbergasted ;D Amazing that you went ahead and did it. I'm amazed that it actually already works with interactive apps, color, everything man. So cool.

Both listener and initiator works great on macOS. Listener works well on Manjaro Linux too, but can't get the server to work there currently, which fails launching the child process for some reason. On the initiator I get:

```
[2023-02-13 19:49:28] [Verbose] Link <4884d132dbbd00d7f4b729a70620ba55> established with <rnsh.default.8a615b5bd3783a5a3a2493d4c44df95b/2f527b67dac3801d78d00994a1eed4fc>, RTT is 0.004s
[2023-02-13 19:49:28] [Info]    rnsh.rnsh._execute             Establishing link... [MainThread]
Child process error: [Errno 1] Operation not permitted, command: ['zsh']
```

There's also some issues with it not knowing the terminal type when connecting from Linux to macOS, but I guess it's pretty trivial to get these things worked out. I'll have a look at the code later. If there's anything you need in terms of info from different terminal emulators on Linux, just let me know, and I'll collect whatever info you need.

This is too good man :+1:

**Reply by acehoss** [reply]
Thank you! It has been fun to work on and a deep dive into how TTYs work. And there's a lot more nuance to learn.

> Child process error: [Errno 1] Operation not permitted

I haven't tested on Linux myself yet, but in the 0.0.5 release today there's a speculative fix--tweaking one of the two things that happens after the fork but before the child process is launched. It might be tricky to diagnose that because the error message is coming from the other side of the fork operation.

**Comment by sp00ck** [comment]
very dangerous solutions.
I wonder how to harden it.
I think a good idea is to add one-time passwords and just a simple script for additional encryption.
for example <add xor 'string'> <add aes-128 key,vi> <add symetric>
every user can change order and additional password (from file or from string)

**Reply by acehoss** [reply]
I think the biggest danger would be with users misconfiguring it. Otherwise, it provides equivalent security to SSH with public key authentication.

Users are free to write a script that does additional processing before launching the shell and then jailing connections to that script using the `-C` option.

**Reply by markqvist** [reply]
Hi @sp00ck. Please don't take this personally, it is not. But you are talking bullshit right now. Please think before you make claims.

> very dangerous solutions.

No, it's not. The use of asymmetric key-based authentication is one of the most secure and reliable ways of authentication. How do you think `ssh` works?

If you want to make such idiotic claims, you *really* gotta back it up, and describe exactly *how* this is "very dangerous".

To me it seems you have exactly **zero** idea about how this work.

> I wonder how to harden it.

You should "harden" your own understanding first and foremost.

**Reply by srem1** [reply]
You probably didn't understand. It's not about replacing versus adding encryption. Please read any book by Bruce Schneier. You will understand that changing the algorithm or running it 10000 times, for example, has a significant impact.
https://www.schneier.com/books/

**Reply by acehoss** [reply]
You are free to wrap this software in whatever additional security you want to add. The software assumes the level of security provided by Reticulum is sufficient. If Reticulum's encryption and authentication is not sufficient for your application, you will probably need to find a different technology.

**Reply by markqvist** [reply]
> You probably didn't understand. It's not about replacing versus adding encryption. Please read any book by Bruce Schneier. You will understand that changing the algorithm or running it 10000 times, for example, has a significant impact. https://www.schneier.com/books/

If you want to be taken seriously, you're not doing a great job right now. Trumpeting vague and meaningless statements, and then telling people to go read a book by Schneier is one of the best ways to make yourself look like a complete arse.

When you have understood just the basics of how Reticulum works, and how it employs its various cryptographic systems, let's have an  *informed* discussion on subjects as security and primitive selections.

Until then, please shut up.

**Comment by markqvist** [comment]
Hi @acehoss!

I just tried out your latest release, and I can confirm that it now works under Linux, and even on Android! Performance also seems quite a bit better. I simply can't tell you how amazing this is! You deserve a medal and a cold beer my good sir! I'm so impressed with how well you've made this work in so little time. Hats off to you!

On Linux there is a hang when exiting the remote session, and the listener spins one cpu at 100% when the initiator exits. It hangs for a long time, but does exit eventually.

On Android, this exception is thrown at program startup, but it still works fine afterwards (at least as initiator):

```
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/psutil/_common.py", line 399, in wrapper
    return cache[key]
KeyError: (('/proc',), frozenset())

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/psutil/_pslinux.py", line 285, in <module>
    set_scputimes_ntuple("/proc")
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/psutil/_common.py", line 401, in wrapper
    ret = cache[key] = fun(*args, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/psutil/_pslinux.py", line 268, in set_scputimes_ntuple
    with open_binary('%s/stat' % procfs_path) as f:
  File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/psutil/_common.py", line 728, in open_binary
    return open(fname, "rb", buffering=FILE_READ_BUFFER_SIZE)
PermissionError: [Errno 13] Permission denied: '/proc/stat'
```

I should help you debug those issues, but haven't had time for actually looking into them yet. Thought I'd at least let you know though.

**Reply by markqvist** [reply]
Oh, and I forgot to ask, but I'd love to include this as a package in the RNode Bootstrap Console. Would that be fine with you? When you feel like it's "finished enough" for this, it would be a *very* useful tool to have included there too!

**Reply by acehoss** [reply]
>I simply can't tell you how amazing this is! You deserve a medal and a cold beer my good sir! I'm so impressed with how well you've made this work in so little time. Hats off to you!

I really appreciate the kudos--I went down the rabbit hole for my own enjoyment and interest, but as I'm sure you know, it's nice to know someone appreciates and enjoys your work. This thing went from "I'm going to take a crack at this" to a full blown hyperfocus and I'd hate to see the time card at this point. But it was a lot of fun and definitely a huge win for my first big Python project. And I have to say, I dislike Python a lot less after working on this. I pay the bills by working almost exclusively in C#.

>I just tried out your latest release, and I can confirm that it now works under Linux, and even on Android! Performance also seems quite a bit better.

In that recent release I pulled out the base64 stuff, so that's probably where most of the performance gains came from. I'm really glad that blind bug fix for the `Operation not permitted` issue worked out, because that was going to be a rough one to track down otherwise.

>On Linux there is a hang when exiting the remote session, and the listener spins one cpu at 100% when the initiator exits. It hangs for a long time, but does exit eventually.

There was some trickery required to get Python to do a non-blocking read from a file descriptor, and as a result there's a mix between `async` and threading that seems to leave some futures dangling in some cases. This is probably the cause of the hang. It doesn't help that once the connection is established the TTY gets switched to raw mode and you lose the ability to send signals with the keyboard.

>On Android, this exception is thrown at program startup, but it still works fine afterwards (at least as initiator)

I think the `psutil` error must be coming from some code run on import of the module. It's used to find and close all open file descriptors and sockets after the fork (without which the RNS stack seems to crash). I'm sure there's a way to sort that error out, though.

> I should help you debug those issues, but haven't had time for actually looking into them yet. Thought I'd at least let you know though.

All good, we're making progress. Keep the bug reports coming! I pushed a v0.0.8 an hour go or so with a bunch more fixes.

**Reply by acehoss** [reply]
And do feel free to include this in the bootstrap console 😁👍🏻 Probably wait until v0.1.0, but for a project started a week ago, it's far closer than it deserves to be!
